Sources/KSCrashRecording/KSCrashAppStateTracker.m

```diff
@@ -0,0 +1,314 @@
+        KSCrashAppStateTrackerBlockObserver *obs = (KSCrashAppStateTrackerBlockObserver *)observer;
+        obs.block = nil;
+        obs.object = nil;
+        [self _locked_reapObserversOrObject:nil];
+    }
+    
+    // observers added with an object
+    else {
+        [self _locked_reapObserversOrObject:observer];
+    }
+    
+    os_unfair_lock_unlock(&_lock);
+}
+
+- (KSCrashAppTransitionState)transitionState
+{
+    KSCrashAppTransitionState ret;
+    {
+        os_unfair_lock_lock(&_lock);
+        ret = _transitionState;
+        os_unfair_lock_unlock(&_lock);
+    }
+    return ret;
+}
+
+- (void)_setTransitionState:(KSCrashAppTransitionState)transitionState
+{
+    NSArray<id<KSCrashAppStateTrackerObserving>> *observers = nil;
+    {
+        os_unfair_lock_lock(&_lock);
+        if (_transitionState != transitionState) {
+            _transitionState = transitionState;
+            observers = [_observers copy];
+        }
+        os_unfair_lock_unlock(&_lock);
+    }
+    
+    for (id<KSCrashAppStateTrackerObserving> obs in observers) {
+        [obs appStateTracker:self didTransitionToState:transitionState];
+    }
+}
+
+#define OBSERVE(center, name, block) \
+[center addObserverForName:name \
+object:nil \
+queue:nil \
+usingBlock:^(NSNotification *notification)block] \
+
+- (void)start
+{
+    if (_registrations) {
+        return;
+    }
+    
+    __weak typeof(self)weakMe = self;
+    
+    // Register a normal `exit` callback so we don't think it's an OOM.
+    atexit_b(^{
+        [weakMe _setTransitionState:KSCrashAppTransitionStateExiting];
+    });
+    
+#if TARGET_OS_IOS
+    
+    // register all normal lifecycle events
+    // in the future, we could also look at scene lifecycle
+    // events but in reality, we don't actually need to,
+    // it could just give us more granularity.
+    _registrations = @[
+        
+        OBSERVE(_center, UIApplicationDidFinishLaunchingNotification, {
+            [weakMe _setTransitionState:KSCrashAppTransitionStateLaunching];
+        }),
+        OBSERVE(_center, UIApplicationWillEnterForegroundNotification, {
+            [weakMe _setTransitionState:KSCrashAppTransitionStateForegrounding];
+        }),
+        OBSERVE(_center, UIApplicationDidBecomeActiveNotification, {
+            [weakMe _setTransitionState:KSCrashAppTransitionStateActive];
+        }),
+        OBSERVE(_center, UIApplicationWillResignActiveNotification, {
+            [weakMe _setTransitionState:KSCrashAppTransitionStateDeactivating];
+        }),
+        OBSERVE(_center, UIApplicationDidEnterBackgroundNotification, {
+            [weakMe _setTransitionState:KSCrashAppTransitionStateBackground];
+        }),
+        OBSERVE(_center, UIApplicationWillTerminateNotification, {
+            [weakMe _setTransitionState:KSCrashAppTransitionStateTerminating];
+        }),
+    ];
+    
+#else
+    // on other platforms that don't have UIApplication
```

Comment:
Yeah, they have them in the headers, but there's no implementations and I'm getting errors, it's the weirdest thing.

It's nuts because I've definitely used them on those platforms before.

<img width="935" alt="Screenshot 2024-05-22 at 10 57 41 AM" src="https://github.com/kstenerud/KSCrash/assets/221626/453a81a9-b70b-4ce1-b42b-1a4e9f06b6f1">

Reply:
I got this working.